### DISCUSSION TITLE
✨ Raise a warning if /var/lib/dagger is an overlayfs mount inside the engine container

### DISCUSSION THREAD

**Discussion by marcosnils** [dagger/dagger#12446] [Ideas]
### What are you trying to do?

Whenever the engine is provisioned within a container - default strategy by the SDK's - as highlighted in the operator's manual here (https://github.com/dagger/dagger/blob/main/core/docs/d7yxc-operator_manual.md#execution-requirements), it's imperative to set `/var/lib/dagger` as a volume so it doesn't form part of the overlayfs of the container. Otherwise, this will have a severe performance penalty when running pipelines. 

We should raise a warning if we detect that the engine data path is within an overlayfs so users don't get caught by this.

### Why is this important to you?

Given that we currently don't have an automated way to start a custom engine, users are left with the option of following the operator's manual by themselves which is error prone. We should put safeguards in the engine in the meantime until we ship a better API for this.

### How are you currently working around this?

N/A

**Comment by jedevc** [comment]
Ideally, we should also set `/var/lib/dagger` as a `VOLUME` directory, see:
- https://docs.docker.com/engine/reference/builder/#volume
- https://github.com/opencontainers/image-spec/blob/main/config.md#properties.

Runtimes that follow this should then end up mounting `/var/lib/dagger` to a nice volume path (if the user forgets to do this).

This doesn't prevent the issue when doing incredibly nested setups where more manual intervention is necessary, but it would help to avoid a good number of cases, like as discussed on discord: https://discord.com/channels/707636530424053791/1122940615806685296/1208181520431452190.

**Comment by marcosnils** [comment]
> Ideally, we should also set `/var/lib/dagger` as a `VOLUME` directory, see:

yes, totally agree. I don't think we have a public API to set this in dagger yet?

**Comment by jedevc** [comment]
> > Ideally, we should also set `/var/lib/dagger` as a `VOLUME` directory, see:
> 
> yes, totally agree. I don't think we have a public API to set this in dagger yet?

Yeah correct :smile: much bikeshedding to be done here, I think it's actually quite a rare use case.

A proposal: maybe we could have `WithMountedEmpty` or something - when run inside dagger using a `WithExec` we try and have an empty mount, so we could replace our weird cache volumes use: https://github.com/dagger/dagger/blob/872a34f23c46016700298d89a27e3f3ef6815718/internal/mage/engine.go#L433

Then, if this ends up in an export, we convert it to a `VOLUME` command. How does something like that sound?

**Comment by marcosnils** [comment]
> A proposal: maybe we could have `WithMountedEmpty` or something - when run inside dagger using a `WithExec` we try and have an empty mount, so we could replace our weird cache volumes use:

I like this, maybe use `WithMountedCacheEmpty` so it's clear that this API is still part of mount cache / volunes? Otherwise, `WithMountedEmpty` could also be associated to a `tmpfs` or something else I assume.

**Comment by jedevc** [comment]
> A proposal: maybe we could have `WithMountedEmpty` or something - when run inside dagger using a `WithExec` we try and have an empty mount

Note to future-self: this is actually possible today using a super-naive empty directory: https://gist.github.com/jedevc/80394a6cf7f33f6302dd7d44f82896d7
